### PR TITLE
Let Users Attach devices

### PR DESCRIPTION
**Pull request by Ryex**
The VM needs better tools to build state from objects in order to be able to attach devices other than the main IC smoothly

 - [x] Make slots work via a occupant
 - [x] Attach Fields and Slot filed to derived properties of a device (ie. IC LineNumber, slot logic Occupied and quantity, .etc)
     - [x] Double check that all fields have been linked properly + no Refcell borrow panics 
 - [x] Allow devices to set their own reference id
     - [x] id sequence generator need to account for self set ids
     - [x] ability to reassign id
     - [x] errors for id collision
 - [x] take a "flatted" device object and add it to the VM as a correct device 
 - [x] let device id's be specified in hex (either with or without the $)
 - [x] let devices be removed
 
 This work will start getting into the weeds of how the VM works. Also, I may live to regret the `Rc<RefCell<T>>` decision, alternatives, even if they might require a large refactor, are welcome.
 ... that or I need to follow the game's lead and batch device edits as messages to the VM to resolve after an IC is finished stepping... that may be the smarter choice.

 if you would like to help with any of the above, submit PR's to this branch

### CODE REVIEW

**Comment by Ryex** [comment]
@ComputerDruid @Emilgardis Pinging you in case your not watching this PR

This is nearly ready for merge. still need to let device state be persisted through share links

Preview at:
https://attach-devices--ic10emu.netlify.app/

File: ic10emu/data/instruction_help_patches.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
+  "english": {
+    "bapz": "Branch to line c if abs(a) <= max(b * abs(a), float.epsilon * 8)",
+    "bapzal": "Branch to line c if abs(a) <= max(b * abs(a), float.epsilon * 8) and store next line number in ra",
+    "bnaz": "Branch to line c if abs(a) > max b * abs(a), float.epsilon * 8)",
+    "bnazal": "Branch to line c if abs(a) > max b * abs(a), float.epsilon * 8) and store next line number in ra",
```
Comment (ComputerDruid) [inline ic10emu/data/instruction_help_patches.json:6]: missing `(`

File: ic10emu/src/lib.rs [hunk 2]
```diff
@@ -1120,32 +1183,56 @@ impl VM {
...
-                return Err(ICError::ConnectionIndexOutOfRange(connection, conn_len).into());
+            // remove from current network
+            if let Some(net) = net {
+                if let Some(network) = self.networks.get(net) {
+                    // if there is no other connection to this network
+                    if connections.clone().iter().enumerate().all(|(index, conn)| {
+                        if let Connection::CableNetwork { net: other_net, .. } = conn {
+                            !(other_net.is_some_and(|id| id == *net) && index != connection)
+                        } else {
+                            true
```
Comment (ComputerDruid) [inline ic10emu/src/lib.rs:1400]: The nested conditionals here are tricky.

optional idea: what if you instead `.filter()` for connections that _do_ match the network and then see if the `.count() == 1`. That'd cut down on the negations here which I think is why I struggle to read it.

File: ic10emu/src/lib.rs [hunk 3]
```diff
@@ -976,6 +978,35 @@ impl VM {
...
+            if let Ok(mut ic_ref) = ic.try_borrow_mut() {
+                ic_ref.pins.iter_mut().for_each(|pin| {
+                    if pin.is_some_and(|d| d == old_id) {
+                        pin.replace(new_id);
+                    }
+                })
+            }
+        });
+        self.networks.iter().for_each(|(_net_id, net)| {
+            if let Ok(mut net_ref) = net.try_borrow_mut() {
```
Comment (ComputerDruid) [inline ic10emu/src/lib.rs:1152]: When would it already be borrowed? I guess same for pins, above, but failing to update pins isn't as bad as failing to update networks.

**Review by ComputerDruid — COMMENTED** [review]
So far I've only read the first 5 commits, will try to come back and read the rest later tonight

File: ic10emu/src/lib.rs (hunk 2 above)
Comment (Ryex) [reply ic10emu/src/lib.rs:1400]: .. that would be a smart method

**Review by Emilgardis — COMMENTED** [review]
It looks good! Some pointers:

- Adding a device with power only from network 1 seems to make it accessible from the active ic.
- Would be nice to present in-game name and have that also be searchable.
- The field display would benefit from being expandable, or a way to reorder and hide fields.
- id of chip is unchangable

**Review by ComputerDruid — COMMENTED** [review]
Made it through the rest of the commits.

I also tried out the demo link in Firefox, and came up with a handful of UI notes:

- When changing device id or name of a device attached to a pin, the pin UI doesn't update with the new ID or name. It does seem like it stays attached to the pin in the VM, though.
- When adding devices, clicking "add device" starts with the text box focused but with the previous search still there. So if I click "add device" and then start typing, it mashes the two search terms together, which isn't what I want. Probably just auto-selecting the text is the right way to go there, so that typing overwrites it but the previous search is still there if I want to add a bunch of volume pumps.
- New devices added via UI should probably start with all cable networks connected to "Network 1". I kept forgetting to do that and then being confused when the device wasn't in the pin drop-down. I think if I wanted to keep cable networks separate, I'd just want to create more networks with unique names and leave "Network 1" as the "default network".
- When using the search UI, it lagged a bunch on the first letter while I was typing "analyzer". Presumably it was matching a lot of things and populating the UI, but I dunno because as soon as it stopped lagging the rest of my typing filled in and it filtered down to just the 3 I expected. Maybe just a limit on the number of search results would help?
- I wasn't able to test the script I pasted much because I didn't see a way to control the value of readonly logic fields in the UI. So the liquid pipe analyzer I added has a `Temperature` field but I couldn't set it to 273.15 to see if my script would open the radiator.

This is super exciting though, I really like the devices UI in general. It's going to be soooo nice to use once everything's wired up together.

File: www/src/ts/utils.ts [hunk 4]
```diff
@@ -175,5 +175,38 @@ export function parseNumber(s: string): number {
...
+  if (/^%[01]+$/.test(s)) {
+    return parseInt(s.slice(1), 2)
+  }
+  if (/^\$[0-9A-Fa-f]+$/.test(s)) {
+    return parseInt(s.slice(1), 16)
+  }
+  const hex = parseHex(s);
+  if (!isNaN(hex)) {
+    return hex;
+  }
```
[the remark this answers is not in the document]
Comment (Ryex) [reply www/src/ts/utils.ts]: goo catch, what I wanted was to parse as hex if there was hex digits even if it didn't start with `$`

**Comment by Ryex** [comment]
> Made it through the rest of the commits.
> 
> I also tried out the demo link in Firefox, and came up with a handful of UI notes:
> 
>     * When changing device id or name of a device attached to a pin, the pin UI doesn't update with the new ID or name. It does seem like it stays attached to the pin in the VM, though.
> 

This is the Select updating bug I've reported upstream, I'll need to completely re-implement the whole component to fix it myself

>     * When adding devices, clicking "add device" starts with the text box focused but with the previous search still there. So if I click "add device" and then start typing, it mashes the two search terms together, which isn't what I want. Probably just auto-selecting the text is the right way to go there, so that typing overwrites it but the previous search is still there if I want to add a bunch of volume pumps.
> 

changed

>     * New devices added via UI should probably start with all cable networks connected to "Network 1". I kept forgetting to do that and then being confused when the device wasn't in the pin drop-down. I think if I wanted to keep cable networks separate, I'd just want to create more networks with unique names and leave "Network 1" as the "default network".
> 

hmm, that probably is the desired default behavior, changed

>     * When using the search UI, it lagged a bunch on the first letter while I was typing "analyzer". Presumably it was matching a lot of things and populating the UI, but I dunno because as soon as it stopped lagging the rest of my typing filled in and it filtered down to just the 3 I expected. Maybe just a limit on the number of search results would help?
> 

This is mostly a factor of rendering all the search results. the actual search is extremely fast. I bumped the search delay to 500 (half a second) form 200 (1/5 second) after the filter change so hopefully there will be enough time to type more than 2 characters.

>     * I wasn't able to test the script I pasted much because I didn't see a way to control the value of readonly logic fields in the UI. So the liquid pipe analyzer I added has a `Temperature` field but I couldn't set it to 273.15 to see if my script would open the radiator.
> 

a good point, changed.

> 
> This is super exciting though, I really like the devices UI in general. It's going to be soooo nice to use once everything's wired up together.

**Comment by Ryex** [comment]
> It looks good! Some pointers:
> 
>     * Adding a device with power only from network 1 seems to make it accessible from the active ic.

networks should now respect if a connection is data or power only

> 
>     * Would be nice to present in-game name and have that also be searchable.
> 

not sure what you mean by pre4sent the in game name, but I added a fuzzy filter that will filter the device list based on device name ore prefab name

>     * The field display would benefit from being expandable, or a way to reorder and hide fields.
> 

I improved visibility there, hopefuly that helps

>     * id of chip is unchangable

as in the chip in the slot?
yeah slots need a UI upgrade, it's next on the list

**Comment by Emilgardis** [comment]
> not sure what you mean by pre4sent the in game name, but I added a fuzzy filter that will filter the device list based on device name ore prefab name

I mean the title, something like

image

maybe

made hash and name italics and smaller size

**Pull request merged**